docker/base.dockerfile

```diff
@@ -1,43 +1,30 @@
-FROM ubuntu:jammy
-
-ENV NODE_VERSION=16.20.0
-
-# install required packages
-RUN apt-get update && \
-    DEBIAN_FRONTEND=noninteractive apt-get install -y \
-    curl \
-    software-properties-common \
-    gettext \
-    git \
-    git-lfs \
-    psmisc \
-    python3 \
-    python3-pip \
-    python3-sphinx
-
-# add yarn ppa
-RUN curl -sS https://dl.yarnpkg.com/debian/pubkey.gpg | apt-key add -
-RUN echo "deb https://dl.yarnpkg.com/debian/ stable main" | tee /etc/apt/sources.list.d/yarn.list
-
-# install nodejs and yarn
-RUN apt-get update && \
-    ARCH=$(dpkg --print-architecture) && \
-    curl -sSO https://deb.nodesource.com/node_16.x/pool/main/n/nodejs/nodejs_$NODE_VERSION-1nodesource1_$ARCH.deb && \
-    dpkg -i ./nodejs_$NODE_VERSION-1nodesource1_$ARCH.deb && \
-    rm nodejs_$NODE_VERSION-1nodesource1_$ARCH.deb && \
-    apt-get install yarn
-
-RUN git lfs install
-
-# Check if symbolic links exist before creating them
-RUN if [ ! -f /usr/bin/python ]; then ln -s /usr/bin/python3 /usr/bin/python; fi \
- && if [ ! -f /usr/bin/pip ]; then ln -s /usr/bin/pip3 /usr/bin/pip; fi
-
-# copy Kolibri source code into image
-COPY . /kolibri
-
-# do the time-consuming base install commands
-RUN cd /kolibri \
-    && pip3 install -r requirements/dev.txt \
-    && pip3 install -r requirements/test.txt \
-    && yarn install --network-timeout 100000
+# This Dockerfile serves as a base image for Kolibri installations. It installs python, required OS packages, node, npm, yarn and python dependencies.
+
+# Declare python version argument.
+ARG PYTHON_VARIANT="3.9-slim"
+FROM python:${PYTHON_VARIANT}
+
+# Declare node major version argument.
+ARG NODE_MAJOR_VERSION="16"
+
+# Install only the absolutely-required packages to run Kolibri -- Git, Git-LFS, ip, ps and Node (plus npm).
+RUN DEBIAN_FRONTEND=noninteractive apt-get update && apt-get install -y ca-certificates curl gnupg git git-lfs procps iproute2 \
+  && mkdir -p /etc/apt/keyrings \
+  && curl -fsSL https://deb.nodesource.com/gpgkey/nodesource-repo.gpg.key | gpg --dearmor -o /etc/apt/keyrings/nodesource.gpg \
+  && echo "deb [signed-by=/etc/apt/keyrings/nodesource.gpg] https://deb.nodesource.com/node_$NODE_MAJOR_VERSION.x nodistro main" | tee /etc/apt/sources.list.d/nodesource.list \
+  && apt-get update && apt-get install -t nodistro -y nodejs
+
+# Install yarn.
+RUN npm install --global yarn
+
+# Change working directory for the commands that follow.
+WORKDIR /kolibri
```

Comment:
Initially, my thought process was on those lines so I kept the devcontainer dockerfile completely isolated from deployable dockerfiles. But then eventually it turned out that keeping maintainability in mind utilising the base dockerfile would be better so went that route.

Reply:
cc @rtibbles

Reply:
> keeping maintainability in mind utilising the base dockerfile

These changes aren't backwards incompatible though, so you haven't kept any maintainability.

The two images, a dev container image and a deployable image, have some commonalities such as installing python requirements, but overall have different use cases and responsibilities. For instance, a deployable Kolibri image doesn't need node.js (an existing dilemna) because node.js is only needed for building static JS/CSS resources.

Reply:
Yeah, so having a separate dockerfile for devcontainer is better, isn't it? :thinking:

Reply:
We discussed this a while back, and I asked that you think about this holistically considering we have the desire to make running Kolibri in a docker container more well documented and to potentially have a well-defined shippable image. As I said, the two use cases have commonalities in how define and build these docker images but they have different responsibilities. With `base.dockerfile` and `dev.dockerfile` we already have some of that implemented and separated, but it isn't perfect and it isn't adapted for dev containers. Your changes here have integrated dev containers with what we already have, but have targeted much of that to `base` and thus muddled `base` vs `dev`. TL;DR we already have a separate dockerfile for development but your changes here only affect `base.dockerfile`!

Reply:
I totally agree with you on this that the current approach of this PR has put everything in just the `base.dockerfile`.

I think we need to decide on how our shippable docker image for Kolibri will look like so that we can decide whether we should build our devcontainer image on top of that or not. I think I should talk to David to sort this out as soon as we can...?